feat: Add student specific data table

bot190 opened the pull request:
Adds a new student table for student specific data. Currently it tracks the student's Graduation year. This info is displayed in the user table and on a users profile (if its available).

bot190 committed:
feat: Add student specific data table

Adds a new student table for student specific data. Currently it tracks the student's Graduation year. This info is displayed in the user table and on a users profile (if its available).

msoucy commented on signinapp/__init__.py:
```diff
@@ -118,10 +118,9 @@ def init_default_db():
-    display = User.make("display", "display", password="1234",
-                        role=DISPLAY, approved=True)
-    db.session.add_all([admin_user, display])
+    User.make("admin", "admin", password="1234", role=ADMIN, approved=True)
```
Don't forget to change the test users below

msoucy commented on signinapp/auth.py:
```diff
@@ -9,7 +9,8 @@
 from wtforms.validators import DataRequired, EqualTo
 
-from .model import Guardian, Role, ShirtSizes, Subteam, User, db, get_form_ids
+from .model import Guardian, ShirtSizes, Subteam, User, db, get_form_ids, Student
```
Alt shift o to clean up the imports

msoucy commented on signinapp/auth.py:
```diff
@@ -36,7 +37,8 @@ class RegisterForm(FlaskForm):
         "T-Shirt Size", choices=ShirtSizes.get_size_names(), validators=[DataRequired()])
-
+    graduation_year = SelectField("Grade", choices=list(generate_grade_choices().items()),
+                        validators=[DataRequired()])
```
Not sure this part is necessary since it's a drop down menu

msoucy commented on signinapp/auth.py:
```diff
@@ -81,7 +83,7 @@ def register():
         # if this returns a user, then the user already exists in database
         username = form.username.data
-        user = User.from_username(name)
+        user = User.from_username(username)
```
Whoops. Thanks.

msoucy commented on signinapp/model.py:
```diff
@@ -19,7 +19,8 @@
                    normalize_phone_number_for_storage,
-                   normalize_phone_number_from_storage)
+                   normalize_phone_number_from_storage,
+                   generate_grade_choices)
```
Use the import cleaner here too

msoucy commented on signinapp/model.py:
```diff
@@ -140,11 +141,16 @@ class User(UserMixin, db.Model):
     guardians = db.relationship(
         "Guardian", secondary=parent_child_association_table, back_populates="students")
 
+    # Student specific data
```
Technically a list of guardians should probably be student specific data too?

bot190 replied:
Yup. Moved it.

msoucy commented on signinapp/model.py:
```diff
@@ -219,18 +225,24 @@ def make(username: str, name: str, password: str, role: Role, approved=False, su
     def make_guardian(name: str, phone_number: str, email: str, contact_order: int = 1):
         role = Role.from_name("guardian_limited")
+        # Generate a username that *should* be unique.
+        username = f"{name}_{normalize_phone_number_for_storage(phone_number)}"
```
We'll need some way to change the username when they create a full guardian account

bot190 replied:
Yeah, we'll have to handle this as part of the special guardian registration page. (Match user on name and phone number instead of username)

msoucy commented on signinapp/model.py:
```diff
@@ -291,6 +303,33 @@ def get_from(cls, name: str, phone_number: str, email: str, contact_order: int)
+    @property
+    def display_grade(self):
+        grades = generate_grade_choices()
+        return grades[self.graduation_year]
```
Should we have a check here to prevent invalid lookups (i.e. looking at an old database with a student who has graduated)

bot190 replied:
Good point. I'll add a default for "Alumni"

msoucy commented on signinapp/model.py:
```diff
@@ -291,6 +303,33 @@ def get_from(cls, name: str, phone_number: str, email: str, contact_order: int)
+    def make(cls, name: str, password: str, graduation_year: int, **kwargs) -> User:
+        role = Role.from_name("student")
+        student = User.make(name=name, password=password, role=role, **kwargs)
+        student_user_data = Student(user_id=student.id, graduation_year=graduation_year)
```
Is `student.id` populated at this point? I thought it only happened on a commit?

bot190 replied:
Yeah I do a sneaky "db.session.flush()" in `User.make` so that the ID exists.

msoucy commented on signinapp/model.py:
```diff
@@ -291,6 +303,33 @@ def get_from(cls, name: str, phone_number: str, email: str, contact_order: int)
+        return grades[self.graduation_year]
+
+    @classmethod
+    def make(cls, name: str, password: str, graduation_year: int, **kwargs) -> User:
```
Username is required

msoucy commented on signinapp/model.py:
```diff
@@ -291,6 +303,33 @@ def get_from(cls, name: str, phone_number: str, email: str, contact_order: int)
+    user_id = db.Column(db.Integer, db.ForeignKey("users.id"))
+
+    # Extra student information
+    graduation_year = db.Column(db.Integer)
```
Should subteam be specific to students?

bot190 replied:
I think it's worth storing for mentors as well? I suppose we could store it separately for both types?

msoucy replied:
The fact that the majority of user types won't require it, only mentor and student (and lead technically, but that also has student data) makes me think it should live in those tables. Guardians, admin, display, etc all don't use it.

The downside is that it would affect the badge search and user list pages...

I would be fine delaying to 1.1, but if we do change it's a schema change that would move columns around so we have to decide now. If we want to do it later, we'd need it to be in 2.0

msoucy replied:
Not going to make this change at this time

bot190 committed:
chore: Move student <-> guardian link to Student data

Moves  the student part of the student <-> guardian link into the student table to clean things up a bit more. 

Also includes some other review clean-up.

msoucy commented on signinapp/templates/profile.html.jinja2:
```diff
@@ -16,18 +16,21 @@
       </div>
       <div class="col-xl-8 px-4">
-        {% if user.guardians %}
+        {% if user.student_user_data.guardians %}
```
Does this actually work? I would expect that it triggers an attribute error if user data is none, unless Jinja hides it

bot190 replied:
This *does* seem to work surprisingly. Is it worth making it?
```suggestion
        {% if user.student_user_data and user.student_user_data.guardians %}
```

msoucy committed:
Update signinapp/templates/profile.html.jinja2

Co-authored-by: Ben Beauregard <bot190@gmail.com>
